File: java/po/world_simulator/entities/animals/Human.java
package po.world_simulator.entities.animals;

import java.util.Map;
import po.world_simulator.Config;
import po.world_simulator.DataFormat;
import po.world_simulator.SaveParser;
import po.world_simulator.Vector2;
import po.world_simulator.World;
import po.world_simulator.entities.Animal;
import po.world_simulator.entities.Entity;

public class Human extends Animal {
    private int abilityCooldown;
    private static Human instance;

    public Human(Vector2 position, int power, int age, int abilityCooldown) {
        super(power, 4, position, "Human", age, true);
        this.abilityCooldown = abilityCooldown;

        if (instance != null)
            instance.dispose();

        instance = this;
    }

    public Human(Vector2 position) {
        this(position, 5, 0, 0);
    }

    public Human(Map<String, String> entityData){
        this(
            new Vector2(entityData.get("position")),
            Integer.parseInt(entityData.get("power")),
            Integer.parseInt(entityData.get("age")),
            Integer.parseInt(entityData.get("abilityCooldown"))
        );
    }

    public static Human getInstance() {
        return instance;
    }

    @Override
    public void dispose() {
        if (instance == this)
            instance = null;
        super.dispose();
    }

    @Override
    public void action() {
        if (abilityCooldown > 0)
            abilityCooldown--;

        Vector2 moveDirection = Vector2.ZERO;
        int lastInput = World.getInputManager().getLastInput();

        switch (lastInput) {
            case Config.LEFT:
                moveDirection = new Vector2(-1, 0);
                break;
            case Config.RIGHT:
                moveDirection = new Vector2(1, 0);
                break;
            case Config.UP:
                moveDirection = new Vector2(0, -1);
                break;
            case Config.DOWN:
                moveDirection = new Vector2(0, 1);
                break;
            case Config.ABILITY:
                if (abilityCooldown == 0)
                    abilityCooldown = 10;
                break;
        }

        if (!World.getMap().isTileOutOfBounds(this.position.add(moveDirection)) && !moveDirection.equals(Vector2.ZERO))
            move(moveDirection);

        // if(age == 90) cout << "moje plecy!";
        this.age += 1;
    }

    @Override
    public void collision(Entity otherEntity) {
        // ability handle before turn if collision happened
        if (World.getInputManager().getLastInput() == Config.ABILITY && abilityCooldown == 0)
            abilityCooldown = 11;

        if (abilityCooldown > 5) {
            if (otherEntity.getPower() >= this.getPower())
                this.runAway();
            return;
        }

        super.collision(otherEntity);
    }

    public int getAbilityCooldown() {
        return abilityCooldown;
    }

    @Override
    public void kill() {
        if (abilityCooldown > 5)
            return;
        World.getRenderer().addToLog("Human died at age " + this.age);
        super.kill();
    }

    @Override
    public String saveAsString(SaveParser parser) {
        String data = super.saveAsString(parser);
        data += parser.stringifyEntry("abilityCooldown", DataFormat.TYPE_INT, String.valueOf(this.abilityCooldown));
        return data;
    }
}

File: python/po/world_simulator/entities/animals/Human.py
from po.world_simulator.entities.Animal import Animal
from po.world_simulator.Vector2 import Vector2
from po.world_simulator.Config import Config
from po.world_simulator.DataFormat import DataFormat
import po.world_simulator.World as world_module

class Human(Animal):
    instance = None

    def __init__(self, position_or_data, power=5, age=0, abilityCooldown=0):
        if isinstance(position_or_data, dict):
            super().__init__(
                int(position_or_data.get("power", 5)),
                4,
                Vector2(position_or_data.get("position")),
                "Human",
                int(position_or_data.get("age", 0)),
                True
            )
            self.abilityCooldown = int(position_or_data.get("abilityCooldown", 0))
        else:
            super().__init__(power, 4, position_or_data, "Human", age, True)
            self.abilityCooldown = abilityCooldown

        if Human.instance is not None:
            Human.instance.dispose()

        Human.instance = self


    @staticmethod
    def getInstance():
        return Human.instance


    def dispose(self):
        if Human.instance == self:
            Human.instance = None
        super().dispose()


    def action(self):
        if self.abilityCooldown > 0:
            self.abilityCooldown -= 1

        moveDirection = Vector2.ZERO
        lastInput = world_module.World.getInputManager().getLastInput()

        if lastInput == Config.LEFT:
            moveDirection = Vector2(-1, 0)
        elif lastInput == Config.RIGHT:
            moveDirection = Vector2(1, 0)
        elif lastInput == Config.UP:
            moveDirection = Vector2(0, -1)
        elif lastInput == Config.DOWN:
            moveDirection = Vector2(0, 1)
        elif lastInput == Config.ABILITY:
            if self.abilityCooldown == 0:
                self.abilityCooldown = 10

        if not world_module.World.getMap().isTileOutOfBounds(self.position.add(moveDirection)) and moveDirection != Vector2.ZERO:
            self.move(moveDirection)

        # if(age == 90) cout << "moje plecy!";
        self.age += 1


    def collision(self, otherEntity):
        # ability handle before turn if collision happened
        if world_module.World.getInputManager().getLastInput() == Config.ABILITY and self.abilityCooldown == 0:
            self.abilityCooldown = 11

        if self.abilityCooldown > 5:
            if otherEntity.getPower() >= self.getPower():
                self.runAway()
            return

        super().collision(otherEntity)


    def getAbilityCooldown(self):
        return self.abilityCooldown


    def kill(self):
        if self.abilityCooldown > 5:
            return
        world_module.World.getRenderer().addToLog("Human died at age " + str(self.age))
        super().kill()


    def saveAsString(self, parser):
        data = super().saveAsString(parser)
        data += parser.stringifyEntry("abilityCooldown", DataFormat.TYPE_INT, str(self.abilityCooldown))
        return data